### ISSUE TITLE
Calculations of years and months wrong due to leap year

### ISSUE THREAD

**Issue by IDisposable** [EvanHahn/HumanizeDuration.js#156]
The constants uses for milliseconds in a year and milliseconds in a month are incorrect because they are computed from the slightly naïve understanding of leap days in the Gregorian calendar.

A leap day occurs every four years. However, because that's actually slightly too often, if the year is divisible by 100 then it *is not* a leap year. *However*, because _that_ would be too infrequent, if the year is divisible by 400 they it *is* a leap year.

Thus the correct decimal representation of the _average_ length of a year is `365.2425` days. Likewise, since the year is 12 exact months, an _average_ length of a month is `30.436875` days.

**Comment by IDisposable** [comment]
For reference https://www.infoplease.com/leap-year-101-next-when-list-days-calendar-years-calculation-last-rules

**Comment by EvanHahn** [comment]
Ah dates, always more complicated than I think.

I think your change in #157 is worthy but suffers from a similar problem as the issue reported in #152—these are arguably breaking changes. And because this library is in maintenance mode, I'm hesitant. However, they are also arguably _bug fixes_, in which case they would be in scope.

What do you think?

**Comment by IDisposable** [comment]
I agree that they are breaking changes if someone is testing the specific output against a specific input, this is evidenced by the fact that I had to make a bunch of little tweaks to the unit tests.

That said, anyone expecting that level of specificity probably has already noticed this was "wrong" and wrote off the package or the use where they were not rounding. MOST people would have likely went to rounding the years (most likely) or if they really cared, are using some library where the exact date is being displayed relatively to _now_.

I think this library is more likely used for durations (not points-in-time) and most durations in the many-month range are not likely to want exactness... and those that _do_ want exactness in the output would like it to be accurate as well.

TL;DR I think it's a bug worth fixing and unlikely to break people that would be surprised by the definition of 365.2425 :)

Thanks for the introspection, I appreciate that in a library author and respect whichever way you go with this library. If you _do_ merge it, I would expect a minor version bump and a note in the README.MD, but I wasn't sure what your release strategy would be... would be happy to make those changes in that PR or another one.

**Comment by EvanHahn** [comment]
I don't think this would warrant a minor version bump. That would be "if new, backwards compatible functionality is introduced to the public API" per [the semver spec](https://semver.org/). To me, it seems like it's either a patch-level change (because it's a bugfix) or a major version bump (because it's a breaking change).

I think there's an argument to be made for either. I don't feel very strongly, but I'd rather do a major version bump because:

1. It's arguably a breaking change. As you said, someone could be testing the specific output, or relying on the output for some other reason.
1. It's the more careful option. If it's even _possibly_ a breaking change, we should bump the major version.
1. There are a couple of little things we could clean up in a major version anyway (like the `gr` language, which is a deprecated alias for `el`).

I'd be willing to do a major version bump for something like that, even if the library is in maintenance mode.

What do you think? (Sorry for the delay in responding!)

**Comment by IDisposable** [comment]
As to semver, sure... major bump would be reasonable as well. I said minor because it isn't a patch, but it's also not an API change... but major version bump is reasonable as well. This discussion is why I didn't do any of the release-activity sort of work :)

TL;DR
Major bump seems reasonable.

**Comment by EvanHahn** [comment]
Did some more thinking about this, finally.

I think there are a few distinct questions at hand:

1. Is this a breaking change?
1. What should the default durations for months and years be, if at all?
1. Which units should be included by default?

The answer to the first question is easy: if we change anything, then it's a breaking change. (I started [a branch for the next major version](https://github.com/EvanHahn/HumanizeDuration.js/pull/176) if you want to follow along.)

The other two questions are trickier.

Wikipedia [provides a list of time units](https://en.wikipedia.org/wiki/Unit_of_time). Milliseconds, seconds, minutes, hours, days, and weeks are constant. Months and years, however, have a range. In other words, it's easy to answer "how long is an hour, in milliseconds?" and impossible to answer "how long is a month, in milliseconds?"

To inform what we should do, I looked at what other folks do.

- [Python's official `timedelta`](https://docs.python.org/3.8/library/datetime.html#datetime.timedelta) only goes up to weeks, presumably because of this ambiguity.
- [Go's official `Duration` stringification](https://golang.org/src/time/time.go?s=22752:22785#L659) only goes up to hours, and they claim "days can be different lengths". I don't know more about this.
- [Rails's `distance_of_time_in_words`](https://github.com/rails/rails/blob/34991a6ae2fc68347c01ea7382fa89004159e019/actionview/lib/action_view/helpers/date_helper.rb#L29-L158) sidesteps this problem with inexact measurements (e.g., "about 6 years").
- [`pretty-ms`](https://github.com/sindresorhus/pretty-ms/blob/540c8540d74e165a6674bbcf73695a98931a6fea/index.js#L57-L60) says that years are 365 days and doesn't support months or weeks, after the maintainer [acknowledged this ambiguity](https://github.com/sindresorhus/pretty-ms/issues/16).
- [millisec](https://github.com/sungwoncho/millisec) doesn't know about weeks, months, or years (though there is [an issue about it](https://github.com/sungwoncho/millisec/issues/2)).

Many others don't deal with this problem at all:

- [Python's unofficial `dateutils.relativedelta`](https://bazaar.launchpad.net/~dateutil/dateutil/trunk/view/head:/dateutil/relativedelta.py) doesn't seem to do humanization at all, and durations for units don't seem to be specified.
- [Moment's `.diff` function](https://momentjs.com/docs/#/displaying/difference/) is date-aware, but HumanizeDuration.js only deals with milliseconds (even so, they had [a bunch of issues with month/year differences](https://github.com/moment/moment/pull/571))

My rough takeaway from the above: if your library deals with exact durations (e.g., "1.7 minutes" instead of "about 2 minutes"), then you either (1) don't deal with months and years (2) treat them carefully.

I'd like to continue to support months and years, because I think they're useful.

Given that, I think HumanizeDuration.js should:

1. Remove `mo` and `y` from the default list. For example, by default, humanizing 400 days should humanize to `57 weeks, 1 day`, not something like `1 year, 1 month, ...`.
1. Allow users to opt-in to `mo` and `y`, and allow them to set the unit duration as they see fit. This is already supported with the `unitMeasures` option.
1. Consider updating the default durations of `mo` and `y`, as is done in #157.
1. Document the above clearly.

If that sounds good, I'll do this in the next major version.

**Comment by mkvlrn** [comment]
Hi! Is this being worked on?

I'm not sure this would be the cause but I'm getting a weird result using a time diff from luxon that results in 31536000000ms (times are just from month/yeah - 05/2021 to 05/2022) and result this lib is giving me is 11 months, instead of 12.

Sorry for maybe necroing an issue, but it is still open!

Thanks.

**Comment by EvanHahn** [comment]
I'm not aware of anyone actively working on this.

I haven't had time to get to it, unfortunately.

**Comment by Antrikshy** [comment]
If this were to be handled, would it be a change [around here](https://github.com/EvanHahn/HumanizeDuration.js/blob/main/humanize-duration.js#L1573-L1582)?

What if there was an initialization parameter to optionally use precise numbers? Would that be too much, making the current numbers pointless?

**Comment by EvanHahn** [comment]
I think the real problem is that this module doesn't know the beginning and end of a duration, only the number of milliseconds. For example, the delta between February 1 and March 1 is different depending on the year.

Because this library is in maintenance mode, I'm not going to address this issue. If someone wants to fork this library and fix this bug, I will link to it in the README.

[`Intl.DurationFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DurationFormat), which is built into modern browsers, may help. (I've added a mention of this in b961e8cace217f77f4481981ec5108065f0de2af.)
